### PR TITLE
Mastra SDK exporter: supporting ancestry attributes during live tracing

### LINKED ISSUE #712: [Integration] Add live tracing support for Mastra
### Problem

See https://github.com/traceroot-ai/traceroot-ts/issues/74

### Proposed Solution

n/a

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Willing to Help Implement?

- [ ] I would like to work/help with this

### PR DESCRIPTION
**Pull request by lqiu03**
Closes [PR712](https://github.com/traceroot-ai/traceroot/issues/712)

## What this does

Mastra agent traces now show up in TraceRoot's Live Traces feed while the agent is still running, not only after it finishes. The visible result matches every other TraceRoot SDK (OTel TS, OpenAI Agents, Vercel AI) — Mastra was the only integration that didn't emit the two ancestry attributes the backend needs for live-tree rendering.

| | Before | After |
| :--- | :--- | :--- |
| **Live feed during a running agent (e.g. 30s run)** | empty until second 30, then full trace appears at once | trace appears within ~1s, fills in as children finish |
| **Final rendered trace** | identical | identical |
| **OTLP payload** | no path / ids_path keys | both on every span |

End-to-end video attached.

## What changed (technical)

Two new attributes on every emitted span:

| Attribute | Type | Value |
| :--- | :--- | :--- |
| `traceroot.span.path` | `string[]` | Span names from root to self, inclusive |
| `traceroot.span.ids_path` | `string[]` | Ancestor span IDs from root to parent, exclusive of self |

Computed via two `Map<spanId, string[]>` populated in `SPAN_STARTED` by looking up the parent's path and appending self, then read in `SPAN_ENDED` to attach as attributes on the OTel `ReadableSpan`.

Backend already reads both keys — no backend changes needed:
* `backend/worker/otel_transform.py:481` uses `path[0]` to name in-flight traces.
* `frontend/ui/src/features/traces/utils/index.ts` (`enrichSpansWithPending`) uses both to draw placeholder ancestor spans for live streams.

## Memory bounds

| Event | Cleanup |
| :--- | :--- |
| **Span ends** | Map entries for that span removed |
| **All spans in a trace end** | Trace state evicted |
| **SPAN_ENDED never arrives for some span** | TTL sweep (5 min idle, only when `traceMap` > 64) clears the trace and its leftover spans |

The TTL sweep is the only departure from `TraceRootSpanProcessor` (PR #71). It guards against an unbounded map under upstream Mastra bugs or process crashes that drop close events.

### End to End Test
**Before fix:**
image
Mastra agent tracing will only appear collectively after a particular run is completed. This is the root of this PR and this screenshot demonstrates so.

**After fix (Live test run in tracer):**
https://github.com/user-attachments/assets/d2c1f56e-4512-4f1c-8a70-609f5e10822f
With this PR's incoming code, a Mastra agent tracing (with the example video) arrives in four seperate turns, as the agent runs live. This expected behavior matches other existing TraceRoot supported SDKs like OpenAI, Anthropic, and others, etc. 

---
## Summary by cubic
Mastra traces now stream live to TraceRoot by emitting `traceroot.span.path` and `traceroot.span.ids_path` on every span. Traces appear within ~1s and fill in as children finish; memory stays bounded via TTL cleanup.

- **New Features**
  - Live tracing: compute ancestry on SPAN_STARTED, attach on SPAN_ENDED; handles out-of-order/orphaned ends, ignores event spans, and clears state on shutdown.
  - Canonical ancestry: normalize parent IDs in `ids_path` to match exported OTLP span IDs.
  - Memory guard: TTL eviction (5 min idle, only when >64 traces) sweeps stale traces and their ancestry entries.
  - Testing seam: `_spanExporter` injection and `tsx --test` suites cover path logic, normalization, and edge cases. No backend changes required.

- **Dependencies**
  - Bump `@traceroot-ai/mastra` to `0.1.1`.
  - Add dev dependency `tsx` and a `npm test` script.
  - Housekeeping: remove `packages/mastra/package-lock.json`; refresh `pnpm-lock.yaml` for `tsx`.

<sup>Written for commit 327dfb9d4f7ae8c90c61129f1f75cafb275493e1. Summary will update on new commits. <a href="https://cubic.dev/pr/traceroot-ai/traceroot-ts/pull/94?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

File: packages/mastra/src/exporter.ts [hunk 1]
```diff
@@ -179,8 +210,30 @@ export class TraceRootExporter extends BaseExporter {
...
+    // Compute ancestry paths using the parent map — same pattern as TraceRootSpanProcessor.
+    // Gate on parentNamePath (not parentSpanId) to keep path/ids_path in sync:
+    // if the parent was never tracked, treat this span as a root rather than emitting
+    // a path with no parent prefix but a non-empty ids_path.
+    const parentSpanId = span.parentSpanId;
+    const parentNamePath = parentSpanId ? this._namePathBySpanId.get(parentSpanId) : undefined;
+    const parentIdsPath = parentSpanId ? this._idsPathBySpanId.get(parentSpanId) : undefined;
+
+    const namePath: string[] = parentNamePath ? [...parentNamePath, span.name] : [span.name];
+    const idsPath: string[] =
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/mastra/src/exporter.ts]: Good catch, normalizing `parendSpanId` into `idsPath`

File: packages/mastra/src/exporter.ts [hunk 2]
```diff
@@ -179,8 +210,30 @@ export class TraceRootExporter extends BaseExporter {
   }
 
   private trackSpanStart(span: AnyExportedSpan): void {
+    this.sweepStaleTracesIfNeeded();
```
[the remark this answers is not in the document]
Comment (lqiu03) [reply packages/mastra/src/exporter.ts:213]: @cubic-dev-ai I want to push back because lastTouched is refreshed on any SPAN_STARTED/SPAN_ENDED for traces, os any Mastra agent with nested spans keep the entry warm. Eviction as you descirbe requires 64+ concurrent traces and a flat >5min spac with zero child events - very atypical of an agentic behavior. I think loosening it trades a rare UX nit for an unbounded leak, so I push back.

File: packages/mastra/src/exporter.ts [hunk 3]
```diff
@@ -63,8 +67,23 @@ const TR_METADATA_PREFIX = 'traceroot.metadata';
...
+  lastTouched: number;
 };
 
+// Traces whose lastTouched is older than this are evicted on the next sweep.
+// Bounds memory under "SPAN_ENDED never delivered" conditions (process crash
+// mid-span upstream, dropped event, integration bug).
+const STALE_TRACE_TTL_MS = 5 * 60 * 1000;
+// Skip the eviction sweep entirely if traceMap is small — saves the linear
+// walk for the common case where everything is closing normally.
+const SWEEP_THRESHOLD = 64;
```
Comment (XinweiHe) [inline packages/mastra/src/exporter.ts:85]: I'm thinking eventually, we should put the constant in a centralized place, s.t. it would be easier to track and manage, but this is fine for now.

**Review by XinweiHe — APPROVED** [review]
